docs/developing/http-apis.md

```diff
@@ -12,48 +13,147 @@ useful for:
-executing arbitrary scripts from the API host.
+```bash
+curl -H "Authorization: Bearer 49Np9sqkYV4g_FpOQk1p0j1yJlvoHrZm9SVhQt7H2-9" https://alpha-api.sandstorm.io/
+```
 
-There are various ways to obtain an API key:
+Sandstorm is responsible for generating the API token. The API token
+is used for both **access control** and **routing a request** to the
+appropriate grain.
 
-* The user can click the key icon in the top bar when they have an app
-  open.
+When a request comes in with a valid API token, Sandstorm sanitizes
+the request, removing the `Cookie` header and the API token, adding
+typical Sandstorm authentication headers like `X-Sandstorm-User-Id`,
+and passes the request to the app. Sandstorm combines it with the app
+package's `bridgeConfig.apiPath` as part of sending the request to the
+grain. Sandstorm sanitizes responses and removes any `Set-Cookie`
+response header.
+
+The API endpoint is set up to allow **cross-origin requests from any
+origin**, which means you can access an API from `XMLHttpRequest` on
+any domain.
+
+An app can **request the generation of an API token**, and a Sandstorm
+user can manually create a valid API token by clicking on the Webkey icon
+in the Sandstorm shell.
+
+## How to generate an API token
+
+There are various ways to obtain an API token:
 
-* The app itself can make a call to
+* The best is via *offer templates*, where the app specifies textual
+  information for the user of the app, and Sandstorm places the token
+  into this template before displaying it to the user.
+
+* The app itself can make a Cap'n Proto RPC call to
   `HackSessionContext.generateApiToken()`. See the [web publishing
   guide](web-publishing.md) for more about how to access
   `HackSessionContext`.
 
-* In the future, we will implement an OAuth flow allowing a third
-  party to initiate a request for access to the user's apps.
+* The user can click the key icon in the top bar when they have an app
+  open.
+
+In the future, we will implement an OAuth flow allowing a third party
+to initiate a request for access to the user's apps.
+
+## Creating an offer template
+
+An _offer template_ is a way for Sandstorm to display an API token to
+the user without the app being able to see the token.
+
+You can see an example by launching [a GitWeb
+demo](https://oasis.sandstorm.io/appdemo/6va4cjamc21j0znf5h5rrgnv0rpyvh1vaxurkrgknefvj0x63ash).
+
+We implement this as an `IFRAME` from the Sandstorm server. The grain
+cannot peek into the element. To fill the `IFRAME` with helpful information
+for the user, including an API token, client-side Javascript in the grain
+provides a template to Sandstorm, and Sandstorm responds with a URL that
+the app can use as the `SRC` of the `IFRAME`.
+
+To create an offer template:
+
+* Create an `IFRAME` element within your page with a memorable ID. For example:
 
-Either approach results in a
+```html
+<iframe width="100%" height="55px" id="offer-iframe">
```

Comment:
Consider adding some `style=` values to hide border, set margin to 0, etc. so the frame looks as inlined as possible.